src/study_zone_ui.cpp

```diff
@@ -0,0 +1,265 @@
+                              ImGuiInputTextFlags_EnterReturnsTrue | ImGuiInputTextFlags_CallbackCharFilter,
+                              filter_skill_input_callback );
+            skill_filter = filter_buffer.data();
+
+            ImGui::SameLine();
+            if( ImGui::Button( _( "Check All" ) ) ) {
+                for( const std::string &npc_name : npc_names ) {
+                    std::set<skill_id> &npc_skills = npc_skill_preferences[npc_name];
+                    for( const skill_id &skill : filtered_skills ) {
+                        npc_skills.insert( skill );
+                    }
+                }
+                changed = true;
+            }
+
+            ImGui::SameLine();
+            if( ImGui::Button( _( "Clear All" ) ) ) {
+                for( const std::string &npc_name : npc_names ) {
+                    std::set<skill_id> &npc_skills = npc_skill_preferences[npc_name];
+                    for( const skill_id &skill : filtered_skills ) {
+                        npc_skills.erase( skill );
+                    }
+                    if( npc_skills.empty() ) {
+                        npc_skill_preferences.erase( npc_name );
+                    }
+                }
+                changed = true;
+            }
+
+            ImGui::SameLine();
+            if( ImGui::Button( _( "Done" ) ) ) {
+                is_open = false;
+            }
+        }
+
+        cataimgui::bounds get_bounds() override {
+            ImVec2 viewport_size = ImGui::GetMainViewport()->Size;
+            float width = std::min( 800.0f, viewport_size.x * 0.8f );
+            float height = viewport_size.y;
+            return { -1.f, -1.f, width, height };
+        }
+
+        void draw_controls() override {
+            std::vector<skill_id> filtered_skills;
+            if( skill_filter.empty() ) {
+                filtered_skills = all_skills;
+            } else {
+                for( const skill_id &skill : all_skills ) {
+                    if( skill->name().find( skill_filter ) != std::string::npos ) {
+                        filtered_skills.push_back( skill );
+                    }
+                }
+            }
+
+            const float footer_height_to_reserve = ImGui::GetStyle().ItemSpacing.y * 2 +
+                                                   ImGui::GetFrameHeightWithSpacing() * 3;
+
+            // inner scroll
+            if( ImGui::BeginChild( "table_scroll_region", ImVec2( 0, -footer_height_to_reserve ), false,
+                                   ImGuiWindowFlags_HorizontalScrollbar ) ) {
+                // Create a table with Skills as rows and npc names as columns
+                if( ImGui::BeginTable( "skill_npc_table", static_cast<int>( npc_names.size() ) + 1,
+                                       ImGuiTableFlags_Borders | ImGuiTableFlags_ScrollX | ImGuiTableFlags_ScrollY |
+                                       ImGuiTableFlags_RowBg ) ) {
```

Comment:
Have you considered using `ImGuiTableFlags_Sortable` so that the user can select some skills on one NPC then sort by that column so that all the checked skills are at the top where they can be easily replicated to other NPCs?